### ISSUE TITLE
brotli: implement encoder

### ISSUE THREAD

**Issue by jpillora** [dsnet/compress#42]
Please close this if/when brotli encoder is implemented. Nice work btw!

**Comment by dsnet** [comment]
Thanks for the issue. I'll definitely try to get a working brotli encoder released soon. I'll give fair warning that the first cut might not perform much better than DEFLATE in terms of it's compression ratio. I plan on doing a iterative approach to implementing brotli, where each iteration improves it's compression performance.

**Comment by jpillora** [comment]
Cool that sounds fine to me +1

**Comment by dsnet** [comment]
Placing this here for reference: mholt/caddy#525

**Comment by blitzprog** [comment]
Has this been abandoned or is there any chance somebody is still working on a native implementation?

**Comment by dsnet** [comment]
I wouldn't say "abandoned", but I'm not actively working on it at the present moment. I have a number of other projects currently occupying my time.

**Comment by blitzprog** [comment]
I see, thank you for the reply!
Would have loved to see a native implementation in Go but I realize it's a very time-consuming task.

**Comment by shurcooL** [comment]
@dsnet In https://github.com/mholt/caddy/issues/525#issuecomment-174615921, you said:

> I'm actually currently working on finishing my own implementation of a RFC 1951 DEFLATE compressor. The components built to implement DEFLATE will make implementing Brotli significantly easier.

I had a question about that. Given that a pure Go DEFLATE decompressor already exists in Go standard library ([`compress/flate`](https://godoc.org/compress/flate)), what's the reason you're building your own? Is it because you have different objectives compared to the std lib one, or another reason?

Can the components of `compress/flate` compressor be used to help implement Brotli compressor, or is there a reason why `compress/flate` isn't helpful for that?

Just looking to understand this better, thanks.

**Comment by dsnet** [comment]
Obviously, the lack of traffic on this repo implies haven't done much in that regard :disappointed:

The reason are the following:
* I already implemented huffman encoding (since it's the same for bzip2 and brotli), so the only other component left for DEFLATE is just LZ77. Implementing LZ77 would allow me to experiment with different algorithms and understand it better.
* There are low level functionality that I would want to be able to do on DEFLATE that I don't think they standard library should ever support. See the [zlib manual](https://www.zlib.net/manual.html) for what features they provide.

**Comment by blitzprog** [comment]
@dsnet I know absolutely nothing about brotli internals but I have a decent knowledge of Go. Is there a chance for someone like me to help you?

**Comment by dsnet** [comment]
@blitzprog, I appreciate your enthusiasm, but implementation compression algorithms really does require specialized knowledge. You're welcome to do research on how LZ77 works and various approaches to compute it.

**Comment by faradaytrs** [comment]
Any progress on this?

**Comment by dmitshur** [comment]
@dsnet FYI, I was [made aware](https://github.com/shurcooL/vfsgen/issues/71#issue-467147181) (thanks to @mappu) that there is a pure Go implementation of a Brotli encoder that was created using a `c2go`-based approach by @andybalholm:

https://groups.google.com/forum/#!msg/golang-nuts/J5MwJnmDpOw/yQacKG2pBwAJ

I haven't had a chance to try it yet, but I wanted to share it with you (and others here).

**Comment by dsnet** [comment]
Thanks for the update @dmitshur! I've been too busy with my actual work to work on this implementation. I'm glad someone else is providing it.
